# **Problem Set 1 - Problem B**

## **Passing requirements**
**All of the following must be satisfied for a passing grade**
- Do the math and correctly answer the Multiple Choice & True or False questions for interpretation
- **At least 80 points needed to get the assignment approved**

**_Please run the cell below before you start the assignment_**

In [1]:
# Import required Python libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

# Folder for the files
files_folder = "files_PS1B_V1_0/"

## **Contribution analysis**

For this problem set, we look at the stressors and environmental impact potentials associated with the production induced by a given external demand. 

### **Production process i)** _(50 points)_

**_Run the cell below to import the required vectors/matrices._**

In [2]:
A_i = pd.read_csv(files_folder + "A_i.csv", index_col = [0])
y_i = pd.read_csv(files_folder + "y_i.csv", index_col = [0])
S_i = pd.read_csv(files_folder + "S_i.csv", index_col = [0])
C_i = pd.read_csv(files_folder + "C_i.csv", index_col = [0])

**_You can display the matrices/vectors in the cell below if you want to understand the production process you are working with._**

In [3]:
display(A_i)
display(y_i)
display(S_i)
display(C_i)

A    B
A  0.03  0.4
B  0.50  0.1

Final demand
A             0
B             3

A     B
CO2  0.6  0.01
CH4  0.2  0.10

CO2     CH4
GWP (kgCO2eq)      1  24.000
PCOP (kgC2H4eq)    0   0.006

#### **1i) Vector of total stressors e** _(10 points)_

Define the identity matrix `I_i` and calculate the determinant `detIA_i` of the matrix (I-A) in the cell under.

In [4]:
I_i = np.identity(2)

detIA_i = np.linalg.det(I_i - A_i)
print(detIA_i)

0.6729999999999999


**_Run the cell under to check the format of your answer(s)_**

In [5]:
assert type(detIA_i) == np.float64, 'The answer(s) must be numeric'

Determine the L matrix that you will name `L_i`. Remember that `L_i` must be a Pandas DataFrame.

In [6]:
L_i_arr=np.linalg.inv(I_i-A_i)
L_i=pd.DataFrame(L_i_arr, index = ['A', 'B'], columns = ['A', 'B'])
display(L_i)

A         B
A  1.337296  0.594354
B  0.742942  1.441308

**_Run the cell under to check the format of your answer(s)_**

In [7]:
assert (L_i.index == ['A', 'B']).all(), 'The index do not correspond'
assert (L_i.columns == ['A', 'B']).all(), 'The columns do not correspond'

Calculate the output vector `x_i` using `y_i`. Rename the column of `x_i` as "Output".

In [8]:
x_i=L_i @ y_i
x_i = x_i.rename(columns = {'Final demand' : 'Output'})

display(x_i)

Output
A  1.783061
B  4.323923

**_Run the cell under to check the format of your answer(s)_**

In [9]:
assert (x_i.columns == ['Output']).all(), 'The columns do not correspond'

Compute the vector of total stressors `e_i`. Rename the column of `e_i` as "Stressors".

In [10]:
e_i=S_i @ x_i
e_i = e_i.rename(columns = {'Output' : 'Stressors'})

display(e_i)

Stressors
CO2   1.113076
CH4   0.789004

**_Run the cell under to check the format of your answer(s)_**

In [11]:
assert (e_i.columns == ['Stressors']).all(), 'The column does not correspond'

**True or False?**

The vector `e_i` represents the total emissions for a final demand of 0 unit of product A and 3 units of product B.

Define your answer as the variable `answer_1i` (e.g. `answer_1i = True` if you think the statement is True and `answer_1i = False` if you think the statement is False).

In [12]:
answer_1i = True

**_Run the cell under to check the format of your answer(s)_**

In [13]:
assert all (type(i) in [bool] for i in [answer_1i]), 'The answer(s) must be a boolean i.e. True or False'

#### **2i) Matrix of process stressors E** _(10 points)_

Compute the matrix of stressors by processes `E_i`.

In [15]:
x_i_marr = np.diag(x_i["Output"])
x_i_m = pd.DataFrame(x_i_marr, index = ['A' , 'B'], columns = ['CO2' , 'CH4'])
display(x_i_m)
display(S_i)

E_i = S_i @ x_i_m
E_i = E_i.rename(columns = {'CO2' : 'A' , 'CH4' : 'B'})
display(E_i)

CO2       CH4
A  1.783061  0.000000
B  0.000000  4.323923

A     B
CO2  0.6  0.01
CH4  0.2  0.10

A         B
CO2  1.069837  0.043239
CH4  0.356612  0.432392

**_Run the cell under to check the format of your answer(s)_**

In [16]:
assert (E_i.index == ['CO2', 'CH4']).all(), 'The index do not correspond'
assert (E_i.columns == ['A', 'B']).all(), 'The columns do not correspond'

**Which one(s) of these statement(s) is/are correct?**

1. There is no final demand of process A.
2. There is no final demand of process B
3. Process A emits CO$_2$ and CH$_4$.
4. Process B emits CO$_2$ and CH$_4$.

Define your answer as the variable `answer_2i` (e.g. `answer_2i = {1}` if you think the correct statement is statement {1} and `answer_2i = {1, 2}` if you think the correct statements are statements {1, 2}).

In [17]:
answer_2i = {1, 3, 4}

**_Run the cell under to check the format of your answer(s)_**

In [18]:
assert all (type(i) in [set] for i in [answer_2i]), 'The answer(s) must be a set of values'

#### **3i) Vector of total impacts d** _(10 points)_

Compute the vector of total impacts `d_i`. Rename the column of `d_i` as "Total impacts".

In [19]:
d_i = C_i @ e_i
d_i = d_i.rename(columns = {'Stressors' : 'Total impacts'})
display(d_i)

Total impacts
GWP (kgCO2eq)        20.049183
PCOP (kgC2H4eq)       0.004734

**_Run the cell under to check the format of your answer(s)_**

In [20]:
assert (d_i.columns == ['Total impacts']).all(), 'The column does not correspond'

**Which one(s) of these statement(s) is/are correct?**

1. The first row represents global warming potential in kg of CH$_4$ equivalents.
2. The second row represents an impact whose unit is kg of ethylene equivalents.
3. Two impacts categories are represented by `d_i`.
4. In this context, PCOP means Potential Cost of Production. 

Define your answer as the variable `answer_3i` (e.g. `answer_3i = {1}` if you think the correct statement is statement {1} and `answer_3i = {1, 2}` if you think the correct statements are statements {1, 2}).

In [21]:
answer_3i = {2, 3}

**_Run the cell under to check the format of your answer(s)_**

In [22]:
assert all (type(i) in [set] for i in [answer_3i]), 'The answer(s) must be a set of values'

#### **4i) Matrix of impacts from each process Dpro** _(10 points)_

Compute the matrix of impacts by processes `Dpro_i`.

In [23]:
Dpro_i = C_i @ E_i

display(Dpro_i)

A          B
GWP (kgCO2eq)    9.628529  10.420654
PCOP (kgC2H4eq)  0.002140   0.002594

**_Run the cell under to check the format of your answer(s)_**

In [24]:
assert (Dpro_i.index == ['GWP (kgCO2eq)', 'PCOP (kgC2H4eq)']).all(), 'The index do not correspond'
assert (Dpro_i.columns == ['A', 'B']).all(), 'The columns do not correspond'

**Which one(s) of these statement(s) is/are correct?**

1. Process A contributes more to GWP than process B.
2. Process B contributes more to GWP than process A.
3. Process A emits around 0.00214 kg SO$_2$-equivalents.
4. Process B emits around 10.42 kg CO$_2$-equivalents.

Define your answer as the variable `answer_4i` (e.g. `answer_4i = {1}` if you think the correct statement is statement {1} and `answer_4i = {1, 2}` if you think the correct statements are statements {1, 2}).

In [25]:
answer_4i = {2, 4}

**_Run the cell under to check the format of your answer(s)_**

In [26]:
assert all (type(i) in [set] for i in [answer_4i]), 'The answer(s) must be a set of values'

#### **5i) Matrix of impacts from each stressor Dstr** _(10 points)_

Compute the matrix of impacts by stressors `Dstr_i`.

In [27]:
e_i_marr = np.diag(e_i["Stressors"])
e_i_m = pd.DataFrame(e_i_marr, index = ['A' , 'B'], columns = ['CO2' , 'CH4'])
display(e_i_m)

e_i_m = e_i_m.rename(index = {'A' : 'CO2' , 'B' : 'CH4'})

Dstr_i = C_i @ e_i_m

display(Dstr_i)

CO2       CH4
A  1.113076  0.000000
B  0.000000  0.789004

CO2        CH4
GWP (kgCO2eq)    1.113076  18.936107
PCOP (kgC2H4eq)  0.000000   0.004734

**_Run the cell under to check the format of your answer(s)_**

In [28]:
assert (Dstr_i.index == ['GWP (kgCO2eq)', 'PCOP (kgC2H4eq)']).all(), 'The index do not correspond'
assert (Dstr_i.columns == ['CO2', 'CH4']).all(), 'The columns do not correspond'

**Which one(s) of these statement(s) is/are correct?**

1. CH$_4$ is responsible for most of the impacts for GWP.
2. CO$_2$ does not contribute to PCOP because there are no CO$_2$ emissions.
3. CH$_4$ creates all the impacts for PCOP.  

Define your answer as the variable `answer_5i` (e.g. `answer_5i = {1}` if you think the correct statement is statement {1} and `answer_5i = {1, 2}` if you think the correct statements are statements {1, 2}).

In [29]:
answer_5i = {1, 3}

**_Run the cell under to check the format of your answer(s)_**

In [30]:
assert all (type(i) in [set] for i in [answer_5i]), 'The answer(s) must be a set of values'

### **Production process ii)** _(50 points)_

**Run the cell below to import the required vectors/matrices.**

In [31]:
A_ii = pd.read_csv(files_folder + "A_ii.csv", index_col = [0])
y_ii = pd.read_csv(files_folder + "y_ii.csv", index_col = [0])
S_ii = pd.read_csv(files_folder + "S_ii.csv", index_col = [0])
C_ii = pd.read_csv(files_folder + "C_ii.csv", index_col = [0])

**_You can display the matrices/vectors in the cell below if you want to understand the production process you are working with._**

In [32]:
display(A_ii)
display(y_ii)
display(S_ii)
display(C_ii)

A    B
A  0.03  0.4
B  0.50  0.1

Final demand
A             4
B             0

A    B
NOx  0.8  0.4
SO2  0.2  0.7

NOx  SO2
EP (kgPO3-/4eq)  0.1    0
AP (kgSO2eq)     0.3    1

#### **1ii) Vector of total stressors e** _(10 points)_

Define the identity matrix `I_ii` and calculate the determinant `detIA_ii` in the cell under.

In [33]:
I_ii = np.identity(2)

detIA_ii = np.linalg.det(I_ii - A_ii)
print(detIA_ii)

0.6729999999999999


**_Run the cell under to check the format of your answer(s)_**

In [34]:
assert type(detIA_ii) == np.float64, 'The answer(s) must be numeric'

Determine the L matrix that you will name `L_ii`. Remember that `L_ii` must be a Pandas DataFrame.

In [35]:
L_ii_arr=np.linalg.inv(I_ii-A_ii)
L_ii=pd.DataFrame(L_ii_arr, index = ['A', 'B'], columns = ['A', 'B'])
display(L_ii)

A         B
A  1.337296  0.594354
B  0.742942  1.441308

**_Run the cell under to check the format of your answer(s)_**

In [36]:
assert (L_ii.index == ['A', 'B']).all(), 'The index do not correspond'
assert (L_ii.columns == ['A', 'B']).all(), 'The columns do not correspond'

Calculate the output vector `x_ii` using `y_ii`. Rename the column of `x_ii` as "Output".

In [37]:
x_ii=L_ii @ y_ii
x_ii = x_ii.rename(columns = {'Final demand' : 'Output'})

display(x_ii)

Output
A  5.349183
B  2.971768

**_Run the cell under to check the format of your answer(s)_**

In [38]:
assert (x_ii.columns == ['Output']).all(), 'The columns do not correspond'

Compute the vector of total stressors `e_ii`. Rename the column of `e_ii` as "Stressors".

In [39]:
e_ii=S_ii @ x_ii
e_ii = e_ii.rename(columns = {'Output' : 'Stressors'})

display(e_ii)

Stressors
NOx   5.468053
SO2   3.150074

**_Run the cell under to check the format of your answer(s)_**

In [40]:
assert (e_ii.columns == ['Stressors']).all(), 'The column does not correspond'

**True or False?**

There is no output of product B.

Define your answer as the variable `answer_1ii` (e.g. `answer_1ii = True` if you think the statement is True and `answer_1ii = False` if you think the statement is False).

In [41]:
answer_1ii = False

**_Run the cell under to check the format of your answer(s)_**

In [42]:
assert all (type(i) in [bool] for i in [answer_1ii]), 'The answer(s) must be a boolean i.e. True or False'

#### **2ii) Matrix of process stressors E** _(10 points)_

Compute the matrix of stressors by processes `E_ii`.

In [43]:
x_ii_marr = np.diag(x_ii["Output"])
x_ii_m = pd.DataFrame(x_ii_marr, index = ['A' , 'B'], columns = ['NOx' , 'SO2'])
display(x_ii_m)
display(S_ii)

E_ii = S_ii @ x_ii_m
E_ii = E_ii.rename(columns = {'NOx' : 'A' , 'SO2' : 'B'})
display(E_ii)

NOx       SO2
A  5.349183  0.000000
B  0.000000  2.971768

A    B
NOx  0.8  0.4
SO2  0.2  0.7

A         B
NOx  4.279346  1.188707
SO2  1.069837  2.080238

**_Run the cell under to check the format of your answer(s)_**

In [44]:
assert (E_ii.index == ['NOx', 'SO2']).all(), 'The index do not correspond'
assert (E_ii.columns == ['A', 'B']).all(), 'The columns do not correspond'

**Which one(s) of these statement(s) is/are correct?**

1. There are more emissions of NO$_x$ than SO$_2$.
2. The sum of the rows of `E_ii` gives the vector `e_ii`.
3. The sum of the columns of `E_ii` gives the vector `e_ii`.

Define your answer as the variable `answer_2ii` (e.g. `answer_2ii = {1}` if you think the correct statement is statement {1} and `answer_2ii = {1, 2}` if you think the correct statements are statements {1, 2}).


In [45]:
answer_2ii = {1, 2}

**_Run the cell under to check the format of your answer(s)_**

In [46]:
assert all (type(i) in [set] for i in [answer_2ii]), 'The answer(s) must be a set of values'

#### **3ii) Vector of total impacts d** _(10 points)_

Compute the vector of total impacts `d_ii`. Rename the column of `d_ii` as "Total impacts".

In [47]:
d_ii = C_ii @ e_ii
d_ii = d_ii.rename(columns = {'Stressors' : 'Total impacts'})
display(d_ii)

Total impacts
EP (kgPO3-/4eq)       0.546805
AP (kgSO2eq)          4.790490

**_Run the cell under to check the format of your answer(s)_**

In [48]:
assert (d_ii.columns == ['Total impacts']).all(), 'The column does not correspond'

**True or False?**

The unit of acidification potential (AP) is kgSO$_2$-equivalents.

Define your answer as the variable `answer_3ii` (e.g. `answer_3ii = True` if you think the statement is True and `answer_3ii = False` if you think the statement is False).

In [49]:
answer_3ii = True

**_Run the cell under to check the format of your answer(s)_**

In [50]:
assert all (type(i) in [bool] for i in [answer_3ii]), 'The answer(s) must be a boolean i.e. True or False'

#### **4ii) Matrix of impacts from each process Dpro** _(10 points)_

Compute the matrix of impacts by processes `Dpro_ii`.

In [51]:
Dpro_ii = C_ii @ E_ii

display(Dpro_ii)

A         B
EP (kgPO3-/4eq)  0.427935  0.118871
AP (kgSO2eq)     2.353640  2.436850

**_Run the cell under to check the format of your answer(s)_**

In [52]:
assert (Dpro_ii.index == ['EP (kgPO3-/4eq)', 'AP (kgSO2eq)']).all(), 'The index do not correspond'
assert (Dpro_ii.columns == ['A', 'B']).all(), 'The columns do not correspond'

**True or False?**

Dpro represents the environmental impacts broken down on processes.

Define your answer as the variable `answer_4iia` (e.g. `answer_4iia = True` if you think the statement is True and `answer_4iia = False` if you think the statement is False).

In [53]:
answer_4iia = True

**_Run the cell under to check the format of your answer(s)_**

In [54]:
assert all (type(i) in [bool] for i in [answer_4iia]), 'The answer(s) must be a boolean i.e. True or False'

**Which one(s) of these statement(s) is/are correct?**

Why does process B causes environmental impact even though its final demand is 0?
1. Because process B was taken in the calculations even though there is no final demand for B, we should delete process B.
2. Because process A requires a certain amount of intermediate demand from process B.
3. Because process B is part of the production process and it should be considered even if there is no final demand for it.

Define your answer as the variable `answer_4iib` (e.g. `answer_4iib = {1}` if you think the correct statement is statement {1} and `answer_4iib = {1, 2}` if you think the correct statements are statements {1, 2}).

In [55]:
answer_4iib = {2, 3}

**_Run the cell under to check the format of your answer(s)_**

In [56]:
assert all (type(i) in [set] for i in [answer_4iib]), 'The answer(s) must be a set of values'

#### **5ii) Matrix of impacts from each stressor Dstr** _(10 points)_

Compute the matrix of impacts by stressors `Dstr_ii`.

In [59]:
e_ii_marr = np.diag(e_ii["Stressors"])
e_ii_m = pd.DataFrame(e_ii_marr, index = ['A' , 'B'], columns = ['NOx' , 'SO2'])
display(e_ii_m)

e_ii_m = e_ii_m.rename(index = {'A' : 'NOx' , 'B' : 'SO2'})

Dstr_ii = C_ii @ e_ii_m

display(Dstr_ii)

NOx       SO2
A  5.468053  0.000000
B  0.000000  3.150074

NOx       SO2
EP (kgPO3-/4eq)  0.546805  0.000000
AP (kgSO2eq)     1.640416  3.150074

**_Run the cell under to check the format of your answer(s)_**

In [60]:
assert (Dstr_ii.index == ['EP (kgPO3-/4eq)', 'AP (kgSO2eq)']).all(), 'The index do not correspond'
assert (Dstr_ii.columns == ['NOx', 'SO2']).all(), 'The columns do not correspond'

**Which one(s) of these statement(s) is/are correct?**

1. Summing the columns of `Dpro_ii` or the columns of `Dstr_ii` gives the same result.
2. Summing the rows of `Dpro_ii` or the rows of `Dstr_ii` gives the same result.
3. SO$_2$ contributes only to AP.
4. The characterization matrix `C` is essential in calculating `Dpro_ii` and `Dstr_ii`.

Define your answer as the variable `answer_5ii` (e.g. `answer_5ii = {1}` if you think the correct statement is statement {1} and `answer_5ii = {1, 2}` if you think the correct statements are statements {1, 2}).

In [61]:
answer_5ii = {2, 3, 4}

**_Run the cell under to check the format of your answer(s)_**

In [62]:
assert all (type(i) in [set] for i in [answer_5ii]), 'The answer(s) must be a set of values'